### PR TITLE
Relocatable rpm symlink fix

### PR DESCRIPTION
**Pull request by fsat**
When creating relocatable RPM packages, the symlink needs to take into account the relocated file paths.

RPM does not provide out of the box support for this, hence the pull request to manage symlinks as part of `%post` and `%postun` scriptlet. This will ensure the symlink will be created and removed as part of post-install and post-uninstall step.

When creating the symlink, both the source path and the symlink itself need to take into account the relocated path.

### CODE REVIEW

**Comment by fsat** [comment]
@huntc and @muuki88 - can I get a review please?

File: src/main/scala/com/typesafe/sbt/packager/archetypes/JavaServerApplication.scala [hunk 1]
```diff
@@ -284,6 +284,8 @@ object JavaServerAppPackaging extends AutoPlugin {
     Some(script)
   }
 
+  protected def makePostInstallScript(): Option[String] = ???
+
```
Comment (huntc) [inline src/main/scala/com/typesafe/sbt/packager/archetypes/JavaServerApplication.scala]: Why was this removed?

File: src/main/scala/com/typesafe/sbt/packager/linux/LinuxPlugin.scala [hunk 2]
```diff
@@ -249,4 +249,7 @@ object LinuxPlugin extends AutoPlugin {
     )
   }
 
+  final def chdir(installLocation: String, packageName: String): String =
+    s"$installLocation/$packageName"
+
```
Comment (huntc) [inline src/main/scala/com/typesafe/sbt/packager/linux/LinuxPlugin.scala]: I see little value in the addition of this function. It is another level of indirection and doesn't actually change the dir. :-)

Comment (huntc) [inline src/main/scala/com/typesafe/sbt/packager/linux/LinuxPlugin.scala]: Cancel that - I see that you're calling it from elsewhere

**Comment by huntc** [comment]
LGTM

File: src/main/scala/com/typesafe/sbt/packager/archetypes/JavaServerApplication.scala (hunk 1 above)
Comment (fsat) [inline src/main/scala/com/typesafe/sbt/packager/archetypes/JavaServerApplication.scala]: Oops... I thought I've put that in, I'll put it back.

File: src/main/scala/com/typesafe/sbt/packager/linux/LinuxPlugin.scala (hunk 2 above)
Comment (fsat) [inline src/main/scala/com/typesafe/sbt/packager/linux/LinuxPlugin.scala]: Yes, the `chdir` is actually not change directory - it's the `chdir` variable in the RPM scriptlet which is the actual directory where the application is installed.

**Comment by fsat** [comment]
Manual test successfully done on Redhat 7 VM for both relocated and default installation.

Test is done using ConductR rpms:
* During installation: ensure correct installation path, correct symlink, and application is started successfully.
* During uninstall: ensure application is stopped successfully, files and symlink cleaned up successfully.

File: src/main/scala/com/typesafe/sbt/packager/rpm/RpmMetadata.scala [hunk 3]
```diff
@@ -59,6 +58,47 @@ case class RpmScripts(
...
+    posttrans.fold("")("\n%posttrans\n" + _ + "\n\n")
+
+  def verifyscriptContent(): String =
+    verifyscript.fold("")("\n%verifyscript\n" + _ + "\n\n")
+
+  def preunContent(): String =
+    preun.fold("")("\n%preun\n" + _ + "\n\n")
+
+  def postunContent(tearDownSymlinkScript: Option[String]): String = {
+    val scripts = Seq(postun, tearDownSymlinkScript).collect {
```
Comment (muuki88) [inline src/main/scala/com/typesafe/sbt/packager/rpm/RpmMetadata.scala]: this can be a simple `flatten`, or?

File: src/main/scala/com/typesafe/sbt/packager/rpm/RpmMetadata.scala [hunk 4]
```diff
@@ -59,6 +58,47 @@ case class RpmScripts(
...
   postun: Option[String] = None) {
+
+  def pretransContent(): String =
+    pretrans.fold("")("\n%pretrans\n" + _ + "\n\n")
+
+  def preContent(): String =
+    pre.fold("")("\n%pre\n" + _ + "\n\n")
+
+  def postContent(buildSymlinkScript: Option[String]): String = {
+    val scripts = Seq(post, buildSymlinkScript).collect {
```
Comment (muuki88) [inline src/main/scala/com/typesafe/sbt/packager/rpm/RpmMetadata.scala]: `flatten` ?

File: src/main/scala/com/typesafe/sbt/packager/archetypes/JavaServerApplication.scala (hunk 1 above)
Comment (muuki88) [inline src/main/scala/com/typesafe/sbt/packager/archetypes/JavaServerApplication.scala]: Not implemented? what's missing here?

**Comment by muuki88** [comment]
I'm still confused why the `relocate` command in rpm is necessary as this adds so much complexity. The `defaultLinuxInstallationPath` can't fix this?

Nevertheless code looks okay. And if the relocate feature isn't used, nothing of these bash-magic will make it into the `scriptlets`.

File: src/main/scala/com/typesafe/sbt/packager/archetypes/JavaServerApplication.scala [hunk 5]
```diff
@@ -284,6 +284,8 @@ object JavaServerAppPackaging extends AutoPlugin {
     Some(script)
   }
 
+  protected def makePostInstallScript(): Option[String] = ???
```
Comment (muuki88) [inline src/main/scala/com/typesafe/sbt/packager/archetypes/JavaServerApplication.scala]: Codacy Issue found: [Usage of ??? operator](https://www.codacy.com/app/nepomukseiler/sbt-native-packager/file/1241567461/issues/source?bid=2040925&fileBranchId=2935951#l287)

**Comment by fsat** [comment]
@muuki88, unfortunately the `defaultLinuxInstallationPath` does not fix this. And unfortunately the bash magic will need to be present because the relocation is specified during package install time.

When rpm install is called with relocate option (i.e. `rpm -i <path to my package file> --relocate /usr/share=/opt`), the `defaultLinuxInstallationPath` which is `/usr/share` is now mapped to the new location `/opt`. Internally RPM will look at the `%files` list and move all the files listed under `/usr/share` to `/opt`.

Before this change, the symlink is created during the build time and registered as part of `%files` list. Because of this, specifying the `--relocate` during rpm install will break the symlink.

For example, a symlink from `/usr/bin/myapp` to `/usr/share/myapp/bin/myapp` are created when the package is built. If I were to relocate the `/usr/share` to `/opt`, the `/usr/bin/myapp` symlink will still point to `/usr/share/myapp/bin/myapp` which is no longer valid because all the files are now under `/opt`.

This change will manage the symlink creation as part of post-install and post-uninstall steps using `%post` and `%postun` respectively. The symlink now will point to the correct relocated path. If the package is not relocated (i.e. using the `defaultLinuxInstallationPath`), the `relocateLink` function will leave the link as-is.

The additional complexity introduced in the `%post` and `%postun` scriptlet is unavoidable because rpm does not provide means to handle symlink relocation out of the box.

Hopefully it explains why this change is done.

File: src/sbt-test/jar/classpath-jar/src/main/scala/test/Test.scala [hunk 6]
```diff
@@ -3,17 +3,27 @@ package test
...
+
   val system = ActorSystem("testSystem")
+  import system.dispatcher
+
   val router = system.actorOf(Props[PrintActor])
-  router ! "SUCCESS!!"
-  system.shutdown
+  (router ? "SUCCESS!!").foreach { _ =>
+    system.shutdown
+  }
```
Comment (fsat) [inline src/sbt-test/jar/classpath-jar/src/main/scala/test/Test.scala]: @muuki88 - this is not related to the RPM symlink issue, but I have intermittent failure on this test today, and hence the fix.

The failure is originally caused by the race condition where the actor system may be shutdown before the actor under test has the chance to print the expected message.

**Comment by fsat** [comment]
@muuki88 - if you don't have any further feedback, can you merge this and release another milestone version please?

**Comment by muuki88** [comment]
Released in `1.0.5-RC1`

**Comment by huntc** [comment]
Thanks!

**Comment by fsat** [comment]
Thank you @muuki88!
